Fix test failures if 'future' package is installed

alikins opened the pull request:
##### Issue Type:
- Bugfix Pull Request

The python 'future' module provides a 'builtins' package
to emulate the py3 'builtins' modules. If installed, the
unit tests that reference builtings.**import** fail because
the future 'builtins' is imported and it is missing **import**.

So if/else based on '**builtin**' being in sys.modules for
py2/py3 differences, and only import 'builtins' on py3.

fixes #14996

jpic commented:
lgtm, but perhaps you'd like to split your changes into two commits, one for trailing whitespaces, another for builtins.

alikins committed:
Fix test failures if 'future' package is installed

The python 'future' module provides a 'builtins' package
to emulate the py3 'builtins' modules. If installed, the
unit tests that reference builtins.__import__ fail because
the future 'builtins' is imported and it is missing __import__.

Use ansible.compat.six.moves.builtins instead of importing
'builtins' or __builtin__

Fixes #14996

alikins commented:
Updated without the whitespace changes.

A few style/idiom questions...

The check for '**builtin**' in sys.modules could be replaced with a check of six.PY3 or six.PY2.
This change here was going for 'check for features not versions' rule of thumb, but a version
check here seems reasonable and clearer to me

ie, something like

``` python

from ansible.compat.six import PY2
if PY2:
    builtins = sys.module['__builtin__']
else:
    import builtins
```

Or just use six'es 'builtins' module

``` python
from ansible.compat.six.moves import builtins
```

The latter seems to mostly work, except it causes a unit test to fail on py3 
(units.module_utils.basic.test__log_invocation.TestModuleUtilsBasic) for reasons I haven't tracked
down yet. That test is brand new though...

import of 'ansible.compat.six.moves.builtins'

abadger commented:
Yeah -- use ansible.compat.six.moves.builtins wherever possible.

abadger commented:
If you have things to move to ansible.compat.six.moves.builtins still, feel free to merge them if tests pass with the changes.  If it deviates from that common fix we'll have to evaluate them case by case.

abadger commented:
Using six.PY3 or six.PY2 is also fine.  Feature tests are good (especially if a distro is likely to backport functionality) but in most cases py2 vs py3 is a well-known boundary to coders and something that distros are not crossing.  (The big exception I know of here is TLS/SSL features being backported from python3 to later versions of python2.7)

Looks like test_basic.py has been taken care of but test_action could be switched to use from ansible.compat.six if you want to update and merge.

frenzymadness commented:
Hello @alikins 

If you don't have time to finish this, feel free to ping me and I'll create a new PR to fix this issue.

Have a nice day.
Lumír

alikins commented:
Looking at this again.

Looks like 'future' still breaks test_action.py, albeit in a non obvious way

```
[fedora-25:ansible (devel % u=)]$ pip show future
[fedora-25:ansible (devel % u=)]$ test/runner/ansible-test units --python 2.7 --exclude test/units/modules/ test/units/plugins/action/
Unit test with Python 2.7
==================================================================================================================== test session starts =====================================================================================================================
platform linux2 -- Python 2.7.13, pytest-3.0.4, py-1.4.31, pluggy-0.4.0
rootdir: /home/adrian/src/ansible, inifile: 
plugins: smother-0.1, xdist-1.15.0, mock-1.5.0
collected 27 items 

test/units/plugins/action/test_action.py ...............
test/units/plugins/action/test_raw.py ....
test/units/plugins/action/test_synchronize.py ........

------------------------------------------------------------------------------------ generated xml file: /home/adrian/src/ansible/test/results/junit/python2.7-units.xml -------------------------------------------------------------------------------------
================================================================================================================= 27 passed in 0.71 seconds ==================================================================================================================
[fedora-25:ansible (devel % u=)]$ pip install --user future
Collecting future
  Using cached future-0.16.0.tar.gz
Installing collected packages: future
  Running setup.py install for future ... done
Successfully installed future-0.16.0
[fedora-25:ansible (devel % u=)]$ pip show future
Name: future
Version: 0.16.0
Summary: Clean single-source support for Python 3 and 2
Home-page: https://python-future.org
Author: Ed Schofield
Author-email: ed@pythoncharmers.com
License: MIT
Location: /home/adrian/.local/lib/python2.7/site-packages
Requires: 
[fedora-25:ansible (devel % u=)]$ test/runner/ansible-test units --python 2.7 --exclude test/units/modules/ test/units/plugins/action/
Unit test with Python 2.7
==================================================================================================================== test session starts =====================================================================================================================
platform linux2 -- Python 2.7.13, pytest-3.0.4, py-1.4.31, pluggy-0.4.0
rootdir: /home/adrian/src/ansible, inifile: 
plugins: smother-0.1, xdist-1.15.0, mock-1.5.0
collected 27 items 

test/units/plugins/action/test_action.py .F.............
test/units/plugins/action/test_raw.py ....
test/units/plugins/action/test_synchronize.py ........

------------------------------------------------------------------------------------ generated xml file: /home/adrian/src/ansible/test/results/junit/python2.7-units.xml -------------------------------------------------------------------------------------
================================================================================================================== short test summary info ===================================================================================================================
FAIL test/units/plugins/action/test_action.py::TestActionBase::test_action_base__configure_module
========================================================================================================================== FAILURES ==========================================================================================================================
_____________________________________________________________________________________________________ TestActionBase.test_action_base__configure_module ______________________________________________________________________________________________________

self = <units.plugins.action.test_action.TestActionBase testMethod=test_action_base__configure_module>

    def test_action_base__configure_module(self):
        fake_loader = DictDataLoader({
        })
    
        # create our fake task
        mock_task = MagicMock()
        mock_task.action = "copy"
    
        # create a mock connection, so we don't actually try and connect to things
        mock_connection = MagicMock()
    
        # create a mock shared loader object
        def mock_find_plugin(name, options):
            if name == 'badmodule':
                return None
            elif '.ps1' in options:
                return '/fake/path/to/%s.ps1' % name
            else:
                return '/fake/path/to/%s' % name
    
        mock_module_loader = MagicMock()
        mock_module_loader.find_plugin.side_effect = mock_find_plugin
        mock_shared_obj_loader = MagicMock()
        mock_shared_obj_loader.module_loader = mock_module_loader
    
        # we're using a real play context here
        play_context = PlayContext()
    
        # our test class
        action_base = DerivedActionBase(
            task=mock_task,
            connection=mock_connection,
            play_context=play_context,
            loader=fake_loader,
            templar=None,
            shared_loader_obj=mock_shared_obj_loader,
        )
    
        # test python module formatting
        with patch.object(builtins, 'open', mock_open(read_data=to_bytes(python_module_replacers.strip(), encoding='utf-8'))):
            with patch.object(os, 'rename'):
                mock_task.args = dict(a=1, foo='fö〩')
                mock_connection.module_implementation_preferences = ('',)
>               (style, shebang, data, path) = action_base._configure_module(mock_task.action, mock_task.args)

test/units/plugins/action/test_action.py:135: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
lib/ansible/plugins/action/__init__.py:160: in _configure_module
    task_vars=task_vars, module_compression=self._play_context.module_compression)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

module_name = 'copy', module_path = '/fake/path/to/copy', module_args = {'a': 1, 'foo': 'f\xc3\xb6\xe3\x80\xa9'}, task_vars = {}, module_compression = 'ZIP_DEFLATED'

    def modify_module(module_name, module_path, module_args, task_vars=dict(), module_compression='ZIP_STORED'):
        """
        Used to insert chunks of code into modules before transfer rather than
        doing regular python imports.  This allows for more efficient transfer in
        a non-bootstrapping scenario by not moving extra files over the wire and
        also takes care of embedding arguments in the transferred modules.
    
        This version is done in such a way that local imports can still be
        used in the module code, so IDEs don't have to be aware of what is going on.
    
        Example:
    
        from ansible.module_utils.basic import *
    
           ... will result in the insertion of basic.py into the module
           from the module_utils/ directory in the source tree.
    
        For powershell, there's equivalent conventions like this:
    
        # POWERSHELL_COMMON
    
        which results in the inclusion of the common code from powershell.ps1
    
        """
>       with open(module_path, 'rb') as f:
E       IOError: [Errno 2] No such file or directory: '/fake/path/to/copy'

lib/ansible/executor/module_common.py:810: IOError
============================================================================================================ 1 failed, 26 passed in 0.70 seconds =============================================================================================================
ERROR: Command "pytest -r a --color yes --junit-xml test/results/junit/python2.7-units.xml --ignore=test/units/modules/ test/units/plugins/action/" returned exit status 1.
```

updating to use six.compat.moves.builtins in test_action.py fixes it.

```
[fedora-25:ansible (handle_future_builtins % u+10128-1)]$ pip show future
[fedora-25:ansible (handle_future_builtins % u+10128-1)]$ test/runner/ansible-test units --python 2.7 --exclude test/units/modules/ test/units/plugins/action/
Unit test with Python 2.7
==================================================================================================================== test session starts =====================================================================================================================
platform linux2 -- Python 2.7.13, pytest-3.0.4, py-1.4.31, pluggy-0.4.0
rootdir: /home/adrian/src/ansible, inifile: 
plugins: smother-0.1, xdist-1.15.0, mock-1.5.0
collected 27 items 

test/units/plugins/action/test_action.py ...............
test/units/plugins/action/test_raw.py ....
test/units/plugins/action/test_synchronize.py ........

------------------------------------------------------------------------------------ generated xml file: /home/adrian/src/ansible/test/results/junit/python2.7-units.xml -------------------------------------------------------------------------------------
================================================================================================================= 27 passed in 0.70 seconds ==================================================================================================================
[fedora-25:ansible (handle_future_builtins % u+10128-1)]$ pip install --user future
Collecting future
  Using cached future-0.16.0.tar.gz
Installing collected packages: future
  Running setup.py install for future ... done
Successfully installed future-0.16.0
[fedora-25:ansible (handle_future_builtins % u+10128-1)]$ test/runner/ansible-test units --python 2.7 --exclude test/units/modules/ test/units/plugins/action/
Unit test with Python 2.7
==================================================================================================================== test session starts =====================================================================================================================
platform linux2 -- Python 2.7.13, pytest-3.0.4, py-1.4.31, pluggy-0.4.0
rootdir: /home/adrian/src/ansible, inifile: 
plugins: smother-0.1, xdist-1.15.0, mock-1.5.0
collected 27 items 

test/units/plugins/action/test_action.py ...............
test/units/plugins/action/test_raw.py ....
test/units/plugins/action/test_synchronize.py ........

------------------------------------------------------------------------------------ generated xml file: /home/adrian/src/ansible/test/results/junit/python2.7-units.xml -------------------------------------------------------------------------------------
================================================================================================================= 27 passed in 0.71 seconds ==================================================================================================================
[fedora-25:ansible (handle_future_builtins % u+10128-1)]$ 
```
